### PR TITLE
First part of setting up for height to normal support

### PR DESCRIPTION
**Pull request by bernardkwok**
- Add in the ability to append suffix to inputs and outputs
- Add in code to produce upstream calls using additional suffixes
- Add in test Sobel filter.

### CODE REVIEW

File: source/MaterialXGenShader/SgNode.cpp [hunk 1]
```diff
@@ -804,6 +804,47 @@ SgNode* SgNodeGraph::addNode(const Node& node, ShaderGenerator& shadergen)
     return newNode.get();
 }
 
+
+void SgNodeContext::addInputSuffix(SgInput* input, const string& suffix)
```
Comment (bernardkwok) [inline source/MaterialXGenShader/SgNode.cpp]: Add to the context the ability to set and remove suffixes for inputs and outputs.

File: source/MaterialXGenShader/ShaderGenerator.cpp [hunk 2]
```diff
@@ -130,14 +130,31 @@ void ShaderGenerator::emitInput(const SgInput* input, Shader &shader) const
     {
         shader.addStr(_syntax->getDefaultValue(input->type));
     }
+
+    // Look for any additional suffix to append
+    string suffix;
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ShaderGenerator.cpp]: Append input suffix if any.

File: source/MaterialXGenShader/ShaderGenerator.cpp [hunk 3]
```diff
@@ -130,14 +130,31 @@ void ShaderGenerator::emitInput(const SgInput* input, Shader &shader) const
...
+        shader.addStr(suffix);
+    }
 }
 
-void ShaderGenerator::emitOutput(const SgOutput* output, bool includeType, bool assignDefault, Shader& shader) const
+void ShaderGenerator::emitOutput(const SgNodeContext& context, const SgOutput* output, bool includeType, bool assignDefault, Shader& shader) const
 {
     shader.addStr(includeType ? _syntax->getTypeName(output->type) + " " + output->name : output->name);
+
+    // Look for any additional suffix to append
```
Comment (bernardkwok) [inline source/MaterialXGenShader/ShaderGenerator.cpp]: Add output suffix if any.

File: source/MaterialXTest/ShaderValid.cpp [hunk 4]
```diff
@@ -512,9 +512,14 @@ static void runOSLValidation(const std::string& shaderName, mx::TypedElementPtr
             std::string elementType(element->getType());
             mx::string outputName = element->getName();
 
-            const std::string SURFACE_SHADER("surfaceshader");
+            std::set<std::string> colorClosures =
```
Comment (bernardkwok) [inline source/MaterialXTest/ShaderValid.cpp]: Fix for recent addition of closure tests. Any of these are aliases (pre-processor defines) to a "closure color" in OSL.

File: source/MaterialXGenGlsl/Nodes/HeightFieldToNormal.cpp [hunk 5]
```diff
@@ -0,0 +1,197 @@
+#include <MaterialXGenGlsl/Nodes/HeightFieldToNormal.h>
```
Comment (niklasharrysson) [inline source/MaterialXGenGlsl/Nodes/HeightFieldToNormal.cpp]: For consistency I think the file should be named HeightToNormalGlsl.h/cpp. The convention used for the other nodes is [NodeName]Glsl.

File: source/MaterialXGenGlsl/Nodes/HeightFieldToNormal.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        GlslShaderGenerator shadergen = static_cast<GlslShaderGenerator&>(shadergen_);
+
+        BEGIN_SHADER_STAGE(shader, HwShader::PIXEL_STAGE)
+
+        // Emit function signature. Need to add filter width and filter height
+        shader.addLine("vec3 IM_heighttonormal_vector3_sx_glsl(float K[9], float _scale, float _width, float _height)", false);
+        shader.beginScope();
+
+        // Sobel filter computation. Can change this
+        std::vector<std::string> computeFilter;
```
[the remark this answers is not in the document]
Comment (niklasharrysson) [reply source/MaterialXGenGlsl/Nodes/HeightFieldToNormal.cpp]: Here you could use shader.addBlock(filterSrc) instead, where filterSrc is a string with the complete source block for the filter calculation, including newline characters.

**Pull request merged**